### PR TITLE
Feature/1147 additive secret sharing

### LINKED ISSUE #1147: Implement `Additive Secret Sharing` using node to node communication to replace MP-SPDZ
# Additive Secret Sharing

## Background

Additive secret sharing is a cryptographic technique used to split a secret into multiple shares such that:

- Each share individually provides no information about the secret.
- The secret can be reconstructed when sufficient shares are combined.

Formally, let $s \in Z_{2^m}$ ($m$-bit length) be a secret value that we want to share among $n$ parties. We generate $n-1$ random shares $S_1, S_2, \ldots, [s]_{n-1}$. The $n$-th share $[s]_n$ is calculated as, in a modular arithmetic setting, where $p=2^{m}$, the calculation is:

```
[s]_n = \left(s - \sum_{i=1}^{n-1} [s]_i\right) \mod p 
```

## Protocol

Consider a setting with one server and $n$ nodes, where each node $i$ (for $i > 0$ and $i \leq n$) has a secret key $\text{sk}_i$. The goal is to create a server secret key $\text{sk}_0$ using additive shares. The protocol proceeds as follows:

1. Each node $i$(for $i = 1, 2, \ldots, n$) splits its secret key $\text{sk}_i$ into $n$ additive shares, $[\text{sk}_i]_j$, for $j = 1, 2, \ldots, n$.

2. Each node $i$ sends the share $[\text{sk}_i]_j$ to node $j$ for all $j \neq i$. Node $i$ keeps the share $[\text{sk}_i]_i$ for itself.

3. Each node $i$ computes the sum of the received shares and their own share:

```math
[\text{sk}_0]_i = \sum_{j=1}^{n} [\text{sk}_j]_i
```

4. Each node $i$ sends $[\text{sk}_0]_i$ to the server.

5. The server computes its own share $\text{sk}_0$ as:

```math
\text{sk}_0 = \left(\sum_{i=1}^{n} [ \text{sk}_0 ]_i \right) \mod p
```

- [x] #1183
- [x] #1185
- [x] #1193
- [x] #1184
- [x] #1190
- [x] #1191
- [x] #1197
- [x] #1192

### PR DESCRIPTION
**Pull request by srcansiz**
**PR description**

Closes #1147 #1191 #1192 

Implements additive secret sharing to use in server key generation for Joye-Libert secure aggregation scheme. 

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

File: envs/development/conda/condaenv.pb7hayo2.requirements.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+itk >=5.3.0,<5.4.0
```
Comment (mvesin) [inline envs/development/conda/condaenv.pb7hayo2.requirements.txt:1]: Is it really intended to push this file ?

File: fedbiomed/common/_config.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
+# This file is originally part of Fed-BioMed
+# SPDX-License-Identifier: Apache-2.0
+
+import configparser
```
Comment (mvesin) [inline fedbiomed/common/_config.py:4]: Sorry if I'm missing something, but I don't really understand what is this `_config` for.
- most of the code seems to duplicate `config.py`
- I could not find any use of it (`'config.py` seems to be used)

Same for `fedbiomed/node/_config.py`

File: fedbiomed/common/constants.py [hunk 3]
```diff
@@ -251,7 +251,7 @@ class SecaggElementTypes(_BaseEnum):
         DIFFIE_HELLMAN: one pair of DH key for each node party, public key shared with other node parties
     """
     SERVER_KEY: int = 0
-    DIFFIE_HELLMAN: int = 2
+    DIFFIE_HELLMAN: int = 1
```
Comment (mvesin) [inline fedbiomed/common/constants.py:254]: Don't we have to increase major version of `__messaging_protocol_version__` then ?
The `SecaggRequest` messages will then incompatible values between versions.

(Anyway the protocol major version is increased to 5 in #1211

File: gui/main.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
+"""
```
Comment (mvesin) [inline gui/main.py:1]: Sorry if I'm missing something but is that `gui/main.py` related/wanted in that PR ? Thanks !

File: envs/development/conda/condaenv.pb7hayo2.requirements.txt (hunk 1 above)
Comment (srcansiz) [reply envs/development/conda/condaenv.pb7hayo2.requirements.txt]: sorry, wrong file is pushed

File: fedbiomed/common/secagg/_additive_ss.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+        Raises:
+            FedbiomedValueError: If the secret is not an int or a list of integers.
+        """
+        if not (
+            isinstance(secret, int)
+            or (isinstance(secret, list) and all(isinstance(i, int) for i in secret))
+        ):
+            raise FedbiomedValueError("AdditiveSecret must be an int or a list of int")
+        self.secret = secret
```
Comment (mvesin) [inline fedbiomed/common/secagg/_additive_ss.py:25]:
```suggestion
        self._secret = secret
```

Shouldn't it be a private variable ?

File: fedbiomed/node/secagg/_secagg_setups.py [hunk 6]
```diff
@@ -117,291 +134,228 @@ def element(self) -> Enum:
...
-        # Nevertheless, it finally fails as new context cannot be saved on nodes where it already exists.
-        # Another `secagg_id` should be used or partially existing entry be cleaned from database
         try:
-            self._setup_specific()
+            return self._setup_specific()
         except FedbiomedError as e:
             logger.debug(f"{e}")
-            return self._create_secagg_reply("Can not setup secure aggregation context "
-                                             f"on node for {self._secagg_id}.", False)
-        except Exception as e:
```
[the remark this answers is not in the document]
Comment (srcansiz) [reply fedbiomed/node/secagg/_secagg_setups.py:162]: Hi @mvesin,

Yes it is intentional. A general exception is already caught in the upper layer where setup is executed.

**Comment by ybouilla** [comment]
Hi, why is there twice the same message commit appearing twice? `creating classes as presented in #1184`

**Comment by srcansiz** [comment]
> Hi, why is there twice the same message commit appearing twice? `creating classes as presented in #1184`

Hi @ybouilla,
I don't know but it seems that something went wrong during rebase.

**Review by mvesin — APPROVED** [review]

**Pull request merged**